File: Management/Services/Canvas/CanvasAssignmentService.cs
using CanvasModel.Assignments;
using LocalModels;
using RestSharp;

namespace Management.Services.Canvas;
public interface ICanvasAssignmentService
{
  Task<IEnumerable<CanvasAssignment>> GetAll(ulong courseId);
  Task<ulong> Create(
    ulong canvasCourseId,
    LocalAssignment localAssignment,
    ulong? canvasAssignmentGroupId
  );
  Task Update(
    ulong courseId,
    ulong canvasAssignmentId,
    LocalAssignment localAssignment,
    ulong? canvasAssignmentGroupId
  );
  Task Delete(ulong courseId, ulong assignmentCanvasId, string assignmentName);
  Task CreateRubric(ulong courseId, ulong assignmentCanvasId, LocalAssignment localAssignment);
}
public class CanvasAssignmentService(
  IWebRequestor webRequestor,
  CanvasServiceUtils utils,
  MyLogger<CanvasAssignmentService> logger
  ): ICanvasAssignmentService
{
  private readonly IWebRequestor webRequestor = webRequestor;
  private readonly CanvasServiceUtils utils = utils;
  private readonly MyLogger<CanvasAssignmentService> log = logger;

  public async Task<IEnumerable<CanvasAssignment>> GetAll(ulong courseId)
  {
    var url = $"courses/{courseId}/assignments";
    var request = new RestRequest(url);
    // request.AddParameter("include[]", "overrides");
    var assignmentResponse = await utils.PaginatedRequest<IEnumerable<CanvasAssignment>>(request);
    return assignmentResponse.SelectMany(
      assignments =>
        assignments.Select(
          a => a with { DueAt = a.DueAt?.ToLocalTime(), LockAt = a.LockAt?.ToLocalTime() }
        ).ToArray()
    ).ToArray();
  }

  public async Task<ulong> Create(
    ulong canvasCourseId,
    LocalAssignment localAssignment,
    ulong? canvasAssignmentGroupId
  )
  {
    log.Log($"creating assignment: {localAssignment.Name}");
    var url = $"courses/{canvasCourseId}/assignments";
    var request = new RestRequest(url);
    var body = new
    {
      name = localAssignment.Name,
      submission_types = localAssignment.SubmissionTypes.Select(t => t.ToString()),
      allowed_extensions = localAssignment.AllowedFileUploadExtensions.Select(e => e.ToString()),
      description = localAssignment.GetDescriptionHtml(),
      due_at = localAssignment.DueAt,
      lock_at = localAssignment.LockAt,
      points_possible = localAssignment.PointsPossible,
      assignment_group_id = canvasAssignmentGroupId,
    };
    var bodyObj = new { assignment = body };
    request.AddBody(bodyObj);
    var (canvasAssignment, response) = await webRequestor.PostAsync<CanvasAssignment>(request);
    if (canvasAssignment == null)
      throw new Exception("created canvas assignment was null");


    await CreateRubric(canvasCourseId, canvasAssignment.Id, localAssignment);

    return canvasAssignment.Id;
  }

  public async Task Update(
    ulong courseId,
    ulong canvasAssignmentId,
    LocalAssignment localAssignment,
    ulong? canvasAssignmentGroupId
  )
  {
    log.Log($"updating assignment: {localAssignment.Name}");
    var url = $"courses/{courseId}/assignments/{canvasAssignmentId}";
    var request = new RestRequest(url);
    var body = new
    {
      name = localAssignment.Name,
      submission_types = localAssignment.SubmissionTypes.Select(t => t.ToString()),
      description = localAssignment.GetDescriptionHtml(),
      due_at = localAssignment.DueAt,
      lock_at = localAssignment.LockAt,
      points_possible = localAssignment.PointsPossible,
      assignment_group_id = canvasAssignmentGroupId,
    };

    var bodyObj = new { assignment = body };
    request.AddBody(bodyObj);

    await webRequestor.PutAsync(request);

    await CreateRubric(courseId, canvasAssignmentId, localAssignment);
  }

  public async Task Delete(ulong courseId, ulong assignmentCanvasId, string assignmentName)
  {
    log.Log($"deleting assignment from canvas {assignmentName}");
    var url = $"courses/{courseId}/assignments/{assignmentCanvasId}";
    var request = new RestRequest(url);
    var response = await webRequestor.DeleteAsync(request);
    if (!response.IsSuccessful)
    {
      log.Log(url);
      throw new Exception("Failed to delete assignment");
    }
  }

  public async Task CreateRubric(ulong courseId, ulong assignmentCanvasId, LocalAssignment localAssignment)
  {

    var criterion = new Dictionary<int, object>();

    var i = 0;
    foreach (var rubricItem in localAssignment.Rubric)
    {
      var ratings = new Dictionary<int, object>
      {
        { 0, new { description = "Full Marks", points = rubricItem.Points } },
        { 1, new { description = "No Marks", points = 0 } },
      };
      criterion[i] = new
      {
        description = rubricItem.Label,
        points = rubricItem.Points,
        ratings
      };
      i++;
    }

    // https://canvas.instructure.com/doc/api/rubrics.html#method.rubrics.create
    var body = new
    {
      rubric_association_id = assignmentCanvasId,
      rubric = new
      {
        title = $"Rubric for Assignment: {localAssignment.Name}",
        association_id = assignmentCanvasId,
        association_type = "Assignment",
        use_for_grading = true,
        criteria = criterion,
      },
      rubric_association = new
      {
        association_id = assignmentCanvasId,
        association_type = "Assignment",
        purpose = "grading",
        use_for_grading = true,
      }
    };
    var creationUrl = $"courses/{courseId}/rubrics";
    var rubricCreationRequest = new RestRequest(creationUrl);
    rubricCreationRequest.AddBody(body);
    var (rubricCreationResponse, _) = await webRequestor.PostAsync<CanvasRubricCreationResponse>(
      rubricCreationRequest
    );

    if (rubricCreationResponse == null)
      throw new Exception("failed to create rubric before association");

    var assignmentPointCorrectionBody = new
    {
      assignment = new { points_possible = localAssignment.PointsPossible }
    };
    var adjustmentUrl = $"courses/{courseId}/assignments/{assignmentCanvasId}";
    var pointAdjustmentRequest = new RestRequest(adjustmentUrl);
    pointAdjustmentRequest.AddBody(assignmentPointCorrectionBody);
    var (_, _) = await webRequestor.PutAsync<CanvasAssignment>(pointAdjustmentRequest);
  }
}

File: nextjs/src/services/canvas/canvasAssignmentService.ts
import { CanvasAssignment } from "@/models/canvas/assignments/canvasAssignment";
import { canvasApi, canvasServiceUtils } from "./canvasServiceUtils";
import { LocalAssignment } from "@/models/local/assignment/localAssignment";
import { axiosClient } from "../axiosUtils";
import { markdownToHTMLSafe } from "../htmlMarkdownUtils";
import { CanvasRubricCreationResponse } from "@/models/canvas/assignments/canvasRubricCreationResponse";
import { assignmentPoints } from "@/models/local/assignment/utils/assignmentPointsUtils";

const createRubric = async (
  courseId: number,
  assignmentCanvasId: number,
  localAssignment: LocalAssignment
) => {
  const criterion = localAssignment.rubric.map((rubricItem, i) => ({
    description: rubricItem.label,
    points: rubricItem.points,
    ratings: [
      { description: "Full Marks", points: rubricItem.points },
      { description: "No Marks", points: 0 },
    ],
  }));

  const rubricBody = {
    rubric_association_id: assignmentCanvasId,
    rubric: {
      title: `Rubric for Assignment: ${localAssignment.name}`,
      association_id: assignmentCanvasId,
      association_type: "Assignment",
      use_for_grading: true,
      criteria: criterion,
    },
    rubric_association: {
      association_id: assignmentCanvasId,
      association_type: "Assignment",
      purpose: "grading",
      use_for_grading: true,
    },
  };

  const rubricUrl = `${canvasApi}/courses/${courseId}/rubrics`;
  const rubricResponse = await axiosClient.post<CanvasRubricCreationResponse>(
    rubricUrl,
    rubricBody
  );

  if (!rubricResponse.data) throw new Error("Failed to create rubric");

  const assignmentPointAdjustmentUrl = `${canvasApi}/courses/${courseId}/assignments/${assignmentCanvasId}`;
  const assignmentPointAdjustmentBody = {
    assignment: { points_possible: assignmentPoints(localAssignment) },
  };

  await axiosClient.put(
    assignmentPointAdjustmentUrl,
    assignmentPointAdjustmentBody
  );
};

export const canvasAssignmentService = {
  async getAll(courseId: number): Promise<CanvasAssignment[]> {
    const url = `${canvasApi}/courses/${courseId}/assignments`;
    const { data: assignments } = await axiosClient.get<CanvasAssignment[]>(
      url
    );
    return assignments.map((a) => ({
      ...a,
      due_at: a.due_at ? new Date(a.due_at).toLocaleString() : undefined, // timezones?
      lock_at: a.lock_at ? new Date(a.lock_at).toLocaleString() : undefined, // timezones?
    }));
  },

  async create(
    canvasCourseId: number,
    localAssignment: LocalAssignment,
    canvasAssignmentGroupId?: number
  ) {
    console.log(`Creating assignment: ${localAssignment.name}`);
    const url = `${canvasApi}/courses/${canvasCourseId}/assignments`;
    const body = {
      name: localAssignment.name,
      submission_types: localAssignment.submissionTypes.map((t) =>
        t.toString()
      ),
      allowed_extensions: localAssignment.allowedFileUploadExtensions.map((e) =>
        e.toString()
      ),
      description: markdownToHTMLSafe(localAssignment.description),
      due_at: localAssignment.dueAt,
      lock_at: localAssignment.lockAt,
      points_possible: assignmentPoints(localAssignment),
      assignment_group_id: canvasAssignmentGroupId,
    };

    const response = await axiosClient.post<CanvasAssignment>(url, body);
    const canvasAssignment = response.data;

    if (!canvasAssignment) throw new Error("Created Canvas assignment is null");

    await createRubric(canvasCourseId, canvasAssignment.id, localAssignment);

    return canvasAssignment.id;
  },

  async update(
    courseId: number,
    canvasAssignmentId: number,
    localAssignment: LocalAssignment,
    canvasAssignmentGroupId?: number
  ) {
    console.log(`Updating assignment: ${localAssignment.name}`);
    const url = `${canvasApi}/courses/${courseId}/assignments/${canvasAssignmentId}`;
    const body = {
      name: localAssignment.name,
      submission_types: localAssignment.submissionTypes.map((t) =>
        t.toString()
      ),
      allowed_extensions: localAssignment.allowedFileUploadExtensions.map((e) =>
        e.toString()
      ),
      description: markdownToHTMLSafe(localAssignment.description),
      due_at: localAssignment.dueAt,
      lock_at: localAssignment.lockAt,
      points_possible: assignmentPoints(localAssignment),
      assignment_group_id: canvasAssignmentGroupId,
    };

    await axiosClient.put(url, body);
    await createRubric(courseId, canvasAssignmentId, localAssignment);
  },

  async delete(
    courseId: number,
    assignmentCanvasId: number,
    assignmentName: string
  ): Promise<void> {
    console.log(`Deleting assignment from Canvas: ${assignmentName}`);
    const url = `${canvasApi}/courses/${courseId}/assignments/${assignmentCanvasId}`;
    const response = await axiosClient.delete(url);

    if (!response.status.toString().startsWith("2")) {
      console.error(`Failed to delete assignment: ${assignmentName}`);
      throw new Error("Failed to delete assignment");
    }
  },
};
